src/v/compression/tests/zstd_tests.cc

```diff
@@ -85,6 +87,47 @@ SEASTAR_THREAD_TEST_CASE(stream_zstd_test) {
     }
 }
 
+SEASTAR_THREAD_TEST_CASE(async_stream_zstd_test) {
+    compression::async_stream_zstd fn(default_decompression_size, 1);
+    auto test_sizes = get_test_sizes();
+    for (size_t i : sizes) {
+        iobuf buf = gen(i);
+
+        auto cbuf = fn.compress(buf.share(0, i)).get();
+        auto dbuf = fn.uncompress(std::move(cbuf)).get();
+
+        BOOST_CHECK_EQUAL(dbuf, buf);
+    }
+}
+
+SEASTAR_THREAD_TEST_CASE(async_stream_to_stream_test) {
+    compression::async_stream_zstd fn(default_decompression_size, 1);
+    compression::stream_zstd fn_s;
+    auto test_sizes = get_test_sizes();
+    for (size_t i : sizes) {
+        iobuf buf = gen(i);
+
+        auto cbuf = fn.compress(buf.share(0, i)).get();
+        auto dbuf = fn_s.uncompress(std::move(cbuf));
+
+        BOOST_CHECK_EQUAL(dbuf, buf);
+    }
+}
+
+SEASTAR_THREAD_TEST_CASE(stream_to_async_stream_test) {
+    compression::async_stream_zstd fn(default_decompression_size, 1);
+    compression::stream_zstd fn_s;
+    auto test_sizes = get_test_sizes();
+    for (size_t i : sizes) {
+        iobuf buf = gen(i);
+
+        auto cbuf = fn_s.compress(buf.share(0, i));
+        auto dbuf = fn.uncompress(std::move(cbuf)).get();
+
+        BOOST_CHECK_EQUAL(dbuf, buf);
+    }
+}
+
 SEASTAR_THREAD_TEST_CASE(lz4_block_tests) {
```

Comment:
I think it's important to have at least 1 test which tests [de]compression which is actually interleaved. As far as I can tell we use the blocking `.get()` everywhere (actually it was news to me `.get` was allowed in `SEASTAR_THREAD_TEST`, nice) so each [de]compression will start and finish with no interleaving.

I can imagine plenty of bugs which only crop up when there is interleaving.

Reply:
Will add a test with interleaved compression/decompression operations.